src/chain/chain.rs

```diff
@@ -0,0 +1,331 @@
+use std::marker::PhantomData;
+
+use hal::pso::PipelineStage;
+
+use chain::pass::PassLinks;
+use chain::link::{Acquire, Link, LinkSync, Release};
+use resource::{Access, Layout, Resource, Usage};
+
+/// Unique identifier for resource dependency chain.
+/// Multiple resource can be associated with single chain
+/// if all passes uses them the same way.
+/// Chain id uses marker type so that ids for buffers and images are different.
+#[derive(Derivative)]
+#[derivative(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct ChainId<T: ?Sized>(usize, PhantomData<T>);
+impl<T> ChainId<T> {
+    /// Make new chain id.
+    pub fn new(index: usize) -> Self {
+        ChainId(index, PhantomData)
+    }
+
+    /// Get index value.
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Resource chain for set of resources.
+/// All resources in set are expected to be used only in ways specified by this chain.
+/// Each link of the chain corresponds to some render-pass or similar entity.
+#[derive(Clone, Debug)]
+pub struct Chain<A, L, U, S, W> {
+    usage: U,
+    links: Vec<Option<Link<A, L, S, W>>>,
+}
+
+impl<A, L, U, S, W> Chain<A, L, U, S, W>
+where
+    A: Access,
+    L: Layout,
+    U: Usage,
+{
+    /// Build chain from links defined by passes.
+    pub fn build<P, R, F>(id: ChainId<R>, mut usage: U, passes: P, mut new_semaphore: F) -> Self
+    where
+        P: IntoIterator,
+        P::Item: Borrow<PassLinks<R>>,
+        R: Resource<Access = A, Layout = L, Usage = U>,
+        F: FnMut() -> (S, W),
+    {
+        let mut links: Vec<Option<Link<A, L, S, W>>> = Vec::new();
+
+        // Walk over passes
+        for pass in passes {
+            let pass = pass.borrow();
+            // Collect links from passes.
+            links.push(pass.links.iter().find(|link| link.id == id).map(|link| {
+                usage = usage | link.usage;
+                Link {
+                    queue: pass.queue,
+                    access: link.access,
+                    layout: link.layout,
+                    stages: link.stages,
+                    merged_access: link.access,
+                    merged_layout: link.layout,
+                    merged_stages: link.stages,
+                    acquire: LinkSync::None(Acquire),
+                    release: LinkSync::None(Release),
+                }
+            }));
+        }
+
+        let count = links.len();
+
+        // Walk over all links twice and merge states of compatible sub-chains
+        for index in 0..(count * 2) {
+            let index = index % count;
+            let (before, link_after) = links.split_at_mut(index);
+            let (link, after) = link_after.split_first_mut().unwrap();
+
+            // Skip non-existing
+            let link = if let Some(link) = link.as_mut() {
+                link
+            } else {
+                continue;
+            };
+
+            // Get next existing link
+            if let Some(next) = after
```

Comment:
please explain what is going on here. Unclear to me:
  - the code appears to be going through each link and attempts to merge it with every other link?
  - shouldn't we remove the links after merging?
  - shouldn't we only try to merge with links before the current one?
  - why going twice?

Reply:
I'd like to remove links after merging them, but link is associated with single pass. I'll try to do that in later design iteration.